RapierPhysics: Rapier physics example Improvement

Neosoulink opened the pull request:
> ℹ️ No issue associated with this PR

### Description

I was working on one of my projects based on the `Rapier3D.js`, and I noticed that @mrdoob recently added an example using `Rapier` and `InstancedMesh` which were exactly what I was working on.

Unfortunately, the existing [Rapier addon](https://github.com/mrdoob/three.js/blob/master/examples/jsm/physics/RapierPhysics.js) was a bit restrictive to me at some point (and in the [example page](https://github.com/mrdoob/three.js/blob/master/examples/physics_rapier_instancing.html)  as well).

- We don't have access to the `collider`, `colliderDesc`, ...;
- It comes with its timer, what if I want to branch it to my timer?
- The getter returns only the `rigidBody`, and we're necessarily retrained to use it to get a `mesh`
- The current implementation is mainly designed for `meshes`
- etc...

Because I updated it on my side and where using it, I decided to share my version based and following the original version of the  [RapierPhysics](https://github.com/mrdoob/three.js/blob/master/examples/jsm/physics/RapierPhysics.js).

### What are the changes

- Use the new `addToWorld` instead of `addSceneToWorld` | Example preview.
- Set `OrbitControl` to auto-rotate |  Example preview.
- Add a Window `keyboard` event to apply impulse to objects (to shake) | Example preview.
- Add the`addToWorld()` method to add only one object (and return a bunch of Rapier properties).
- Remove the integrated timer inside `RapierPhysics` but expose the `RapierPhysics.step()` method; 
- Migrate the `RapierPhysics` function implementation to the **class** version.
- **Exposed** more Rapier properties (such `rigidBody`, `collider`, ...). accessible from the `WeakMap` store
- Commented methods (with types support) using `JsDoc`.

#### Here's the updated version coming with this PR

[Screenshot](https://github.com/mrdoob/three.js/assets/44310540/a777b5ec-39bc-4342-9666-b23511d39163)

> 🔖 This is my first contribution here, I've read the **code of conduct** and the **Contribution guidelines** before pushing my changes.
> I may forget something some details, if it's the case, please let me know.
>
> And this is a small change not related to the core of `threejs` but just a small example improvement.
> Hoping that will help someone in the future

Neosoulink committed:
refactor: improve `RapierPhysics` helper accessibility

### Description

- Migrate the **old function** implementation to the **class** version
- `Expose` more Rapier properties (`rigidBody`, `collider`, ...)
  - Store more `RAPIER` props in the `WeakMap`
- Add comments (with types support) using `JsDoc`

Neosoulink committed:
feat: add `keyDown` interactions

### Description

- Use `addToWorld` instead of `addSceneToWorld`
- Add Window `keyboard` event to apply impulse to objects (to shake)

Neosoulink committed:
fix: remove unnecessary check

Neosoulink committed:
feat: disposal methods

Neosoulink committed:
fix: correct `timestep` update

mrdoob commented:
Love the shake!

Neosoulink committed:
fix: correctly remove physic bodies

Neosoulink commented:
I just pushed a small fix for the `removeFromWorld` method, I wasn't **removing** the `collider` and the `rigidBody`, which were accumulating them in the physic world

mrdoob commented on examples/jsm/physics/RapierPhysics.js:
```diff
@@ -1,220 +1,435 @@
-		const radius = parameters.radius !== undefined ? parameters.radius : 1;
-		return RAPIER.ColliderDesc.ball( radius );
-
+export class Physic {
```
I think it should be "Physics" instead of "Physic".

There are a few other instances of "Physic" in other variable names...

Neosoulink committed:
refactor: correct typo

Neosoulink commented:
https://github.com/mrdoob/three.js/assets/44310540/bc87ebb5-dfbf-48ba-b5de-e7e26d10919c

Recently, I used the physics helper to build the above example... Unfortunately in the helper, I was blocked when I wanted to scale `Matrix`s.

I had to update the code a little on my end to be able to `scale` each matrix.

In the `update` method, I added the following line:

```ts
for (let j = 0; j < bodies.length; j++) {
	const physicProperties = bodies[j];

	const position = new Vector3().copy(physicProperties.rigidBody.translation());
	this._quaternion.copy(physicProperties.rigidBody.rotation());
	
	
	const scale = (physicProperties.rigidBody.userData as { scale?: Vector3 })?.scale ?? this._scale; // <--- This line

	this._matrix.compose(position, this._quaternion, scale);
	this._matrix.toArray(array, j * 16);
}
```

I passed a `RAPIER.Vector3` to the `userData` property of the `RigidBody`, otherwise I'm using the common _scale property.
By doing that, I was able to update the scale of individual instance `Matrix`

I think I could help others by being able to control the scale...

> cc: @mrdoob

mrdoob commented:
Live link:
https://raw.githack.com/Neosoulink/three.js/feat/rapier-physic-improvement/examples/index.html?q=rapier#physics_rapier_instancing

mrdoob commented:
Instead of using a keyboard event for the shake do you mind adding a button instead?

![Screenshot 2024-05-13 at 3 03 06 PM](https://github.com/mrdoob/three.js/assets/97088/3e4ac464-c40e-4e7c-b4b6-b521a7cd0f34)

That way it'll work on mobile too.

Neosoulink commented:
> Instead of using a keyboard event for the shake do you mind adding a button instead?
> 
> ![Screenshot 2024-05-13 at 3 03 06 PM](https://private-user-images.githubusercontent.com/97088/329939254-3e4ac464-c40e-4e7c-b4b6-b521a7cd0f34.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU1OTcwMTYsIm5iZiI6MTcxNTU5NjcxNiwicGF0aCI6Ii85NzA4OC8zMjk5MzkyNTQtM2U0YWM0NjQtYzQwZS00ZTdjLWI0YjYtYjUyMWE3Y2QwZjM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTEzVDEwMzgzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxNTZlM2FjZmQ4MWZmYjA1ZmIxOWM1YmUzNDFmNzhiZmJhYjY5ZDU2MmEwNzYzNWE1YzUwNmY4NWM3ZDkzOGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.47QCGjjytgp_NzI0gUj8I8yHwSOZEA2UUk98fdbcdCs)
> 
> That way it'll work on mobile too.

Agree and it makes sense!
I'm updating it

Neosoulink committed:
refactor: make physicsHelper not overwrite matrix scale

Neosoulink committed:
refactor: use `button` to shake

Neosoulink commented:
Hey @mrdoob, I just update it

I added a **button** to shake.

And I updated the `update` method to not fix the `scale` at `1 1 1`
So now we can update the matrix scale and the `physicHelper` will use it.

We can now do the following:
```ts
boxeMatrix.scale( new THREE.Vector3( 2,2,2 ) ); // Selected box matrix
boxes.setMatrixAt( i, mat );
```

And the helper will use the 2 2 2 for that specific **Matrix**

Neosoulink committed:
fix: `deepScan` error resolution

zalo commented:
Don't forget this!   Perhaps do a Ctrl+H to replace `hysic` with `hysics` and `hysicss` with `hysics` to unify the naming scheme?

Neosoulink committed:
fix(typo): correct `physic` to `physics`

Neosoulink commented:
I just realized this PR was delayed from r165 to r167 😆

Neosoulink committed:
fix: correct `RapierPhysics` usage from `xr_ballshooter`

### Description

- Use correct `RapierPhysics` methods
  - `addScene` -> `addSceneToWorld`
  - `setMeshPosition` -> `setObjectPosition`
  - `setMeshVelocity` -> `setObjectVelocity`
- Cut & short `addSceneToWorld` description

Neosoulink commented:
I just realized it was also used in the [xr_ballshooter](https://github.com/mrdoob/three.js/pull/28203/commits/99b9cb01f965b314c657e82462d2fc7341b2e0f8#diff-aefffac21ff190b9e5dbb52abe97f4aa5adc8878343a7086a0d9aab2d470c328) example.
I pushed the fix/update. sorry for not noticing it at the beginning.

cc: @mrdoob @zalo 

https://github.com/user-attachments/assets/20c91037-6067-4e5d-9e45-82fae77fc4d9

Mugen87 commented:
Closing, see https://github.com/mrdoob/three.js/pull/33459#issue-4322070070.
